File: LinkedLists/SwapPairs.java
package LinkedLists;

public class SwapPairs {
  public ListNode swapPairs(ListNode head) {
    if (head == null || head.next == null)
      return head;

    // The new head will always be the second node
    ListNode newHead = head.next;
    ListNode prev = null;
    ListNode current = head;

    while (current != null && current.next != null) {
      if (prev != null) {
        prev.next = current.next;
      }
      ListNode next = current.next.next;
      current.next.next = current;
      current.next = next;
      prev = current;
      current = next;
    }

    return newHead;
  }
}

// Time Complexity: O(n) where n is the number of nodes in the linked list.
// Space Complexity: O(1) since we are only using a constant amount of extra space

File: LinkedLists/swap_pairs.py
from list_node import ListNode


class SwapPairs:
    def swap_pairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head

        # The new head will always be the second node
        newHead = head.next
        prev = None
        current = head
        while current and current.next:
            if prev:
                prev.next = current.next

            prev = current
            next = current.next.next
            current.next.next = current
            current.next = next
            current = next

        return newHead


# Time Complexity: O(n)
# Space Complexity: O(1)
